snr.py

```diff
@@ -0,0 +1,139 @@
+    gain_err = get_shotnoise((args['gain'] * args['ADerr']))
+
+    pixel_terms = args['npix'] * (1 + args['npix'] / args['nb'])
+    detector_noise = args['background'] + args['darkcurrent'] + \
+        readnoise ** 2 + gain_err ** 2
+
+    return args['counts'] / np.sqrt(args['counts'] +
+                                   pixel_terms * detector_noise)
+
+
+def get_unit(arg):
+    """ Converts int/float arguments into quantities with associated units.
+    """
+    if arg == 'counts':
+        return u.electron
+    elif arg in ['npix', 'nb']:
+        return u.pixel
+    elif arg in ['background', 'darkcurrent', 'readnoise']:
+        return u.electron / u.pixel
+    elif arg == 'gain':
+        return u.electron / u.adu
+    elif arg == 'ADerr':
+        return u.adu / u.pixel
+
+
+def get_shotnoise(arg):
+    """ Returns the shot noise (i.e. non-Poissonion noise) in the correct
+    units.
+    """
+    return arg.value ** 2 * np.sqrt(1 * u.electron / u.pixel)
+
+
+def test_units():
+    """ A test to check that snr() returns a quantity given
+    non-quantity arguments.
+    Based on the worked example in "A Handbook to CCD Astronomy",
+    Steven Howell, 2000, pg. 56-57
+    """
+    t = 300
+    readnoise = 5
+    darkcurrent = 22 * t / 60 / 60
+    gain = 5
+    background = 620 * gain
+    nb = 200
+    npix = 1
+    counts = 24013 * gain
+
+    result = snr(counts, npix=npix, background=background,
+                 darkcurrent=darkcurrent, readnoise=readnoise,
+                 gain=gain, nb=nb)
+
+    # the value of the answer given in the text
+    answer = 342 * np.sqrt(1 * u.electron)
+
+    # allow error to be within 1 sigma for this test
+    assert abs(result - answer) < np.sqrt(1 * u.electron)
+
+
+def test_math():
+    """ A test to check that the math in snr() is done correctly.
+    Based on the worked example in "A Handbook to CCD Astronomy",
+    Steven Howell, 2000, pg. 56-57
+    """
+    t = 300 * u.s
+    readnoise = 5 * u.electron
+    darkcurrent = 22 * u.electron / u.pixel / u.hr
+    darkcurrent = (darkcurrent * t).to(u.electron / u.pixel)
+    gain = 5 * u.electron / u.adu
+    background = 620 * u.adu / u.pixel * gain
+    nb = 200 * u.pixel
+    npix = 1 * u.pixel
+    counts = 24013 * u.adu * gain
+
+    result = snr(counts, npix=npix, background=background,
+                 darkcurrent=darkcurrent, readnoise=readnoise,
+                 gain=gain, nb=nb)
+
+    # the value of the answer given in the text
+    answer = 342 * np.sqrt(1 * u.electron)
+
+    # allow error to be within 1 sigma for this test
+    assert abs(result - answer) < np.sqrt(1 * u.electron)
+
+
+def test_bright():
+    """ Test that snr() returns sqrt(counts), the expected value
+    for a bright target.
+    """
+    counts = 25e5
+    answer = np.sqrt(counts * u.electron)
+    result = snr(counts)
+    difference_threshold = 1e-3 * np.sqrt(1 * u.electron)
+    assert abs(result - answer) < difference_threshold
```

Comment:
Not a requirement, but you might find [this function useful](http://docs.astropy.org/en/stable/api/astropy.tests.helper.assert_quantity_allclose.html). [See this astroplan snippet for example](https://github.com/astropy/astroplan/blob/master/astroplan/tests/test_observer.py#L256-L257).

Reply:
Ohh yeah, very useful!